### PR TITLE
Bug Fixes: Python SDK and Feast Core

### LINKED ISSUE #344: Feast serving URL should not be required
Currently the Python SDK requires the Feast Serving URL to be defined when calling the version() method. The failure here should be graceful.

```
ValueErrorTraceback (most recent call last)
<ipython-input-31-b8a3b34abdf1> in <module>
----> 1 client.version()

/opt/notebooks/feast/sdk/python/feast/client.py in version(self)
    105 
    106         self._connect_core()
--> 107         self._connect_serving()
    108 
    109         core_version = ""

/opt/notebooks/feast/sdk/python/feast/client.py in _connect_serving(self, skip_if_connected)
    179         if self.__serving_channel is None:
    180             self.__serving_channel = grpc.insecure_channel(self.serving_url)
--> 181 
    182         try:
    183             grpc.channel_ready_future(self.__serving_channel).result(

ValueError: Please set Feast Serving URL.
```

### LINKED ISSUE #346: Feast should raise an exception when trying to get a feature set that does not exist
## Expected Behavior 
Feast should raise an exception when trying to get a feature set that does not exist, even when the `fail_if_missing` parameter is specified.

## Current Behavior
Feast prints the following error message and does not raise an exception.
```
GetFeatureSet failed with code "StatusCode.UNKNOWN"
```

## Steps to reproduce
```python
from feast import Client
client = Client(core_url, serving_url)
client.get_feature_set("invalid_feature_set_name", fail_if_missing=True)
```

### PR DESCRIPTION
**Pull request by woop**
Fixes
* Fixes a problem where the `apply` feature set method would bump feature set versions if the order of features changed
* Fixed a null pointer exception during `apply` feature set method.
* Fix #346 
* Fix #344

Improvements
* Exceptions are now raised to end users in most cases for Python SDK
* Ordering of feature set results when reading them out through CLI
* Get feature set method now returns a more descriptive error
* Listing feature sets with the Python SDK now prints their names

### CODE REVIEW

File: core/src/main/java/feast/core/model/FeatureSet.java [hunk 1]
```diff
@@ -152,12 +154,56 @@ public FeatureSetSpec toProto() throws InvalidProtocolBufferException {
...
+    // Test amount of fields
+    if (entities.size() != other.entities.size() || features.size() != other.features.size())
+    {
+      return false;
+    }
+
+    // Create a map of all fields in this feature set
+    Map<String, Field> fields = new HashMap<>();
+
+    for (Field e : entities){
```
Comment (zhilingc) [inline core/src/main/java/feast/core/model/FeatureSet.java]: why are we doing validation in the equalTo method?

File: core/src/main/java/feast/core/model/Source.java [hunk 2]
```diff
@@ -173,11 +173,11 @@ public boolean isDefault() {
    * @return boolean equal
    */
   public boolean equalTo(Source other) {
-    if (other.isDefault && isDefault) {
+    if (other.isDefault && isDefault || (type == null && other.type == null)) {
```
Comment (zhilingc) [inline core/src/main/java/feast/core/model/Source.java]: Under what circumstance will this be null?

**Comment by khorshuheng** [comment]
LGTM

**Pull request merged**